Mathlib/LinearAlgebra/FreeModule/PID.lean

```diff
@@ -548,10 +548,121 @@ noncomputable def Submodule.smithNormalForm [Finite ι] (b : Basis ι R M) (N :
       Submodule.comapSubtypeEquivOfLe_apply_coe, Basis.reindex_apply,
       Equiv.toEmbedding_apply, Function.Embedding.trans_apply, Equiv.symm_apply_apply]⟩
 
+section full_rank
+
+variable {N : Submodule R M}
+
+/--
+If `M` is finite free over a PID `R`, then for any submodule `N` of the same rank,
+we can find basis for `M` and `N` with the same indexing such that the inclusion map
+is a square diagonal matrix.
+
+See `Submodule.exists_smith_normal_form_of_rank_eq` for a version that states the
+existence of the basis.
+-/
+noncomputable def Submodule.smithNormalFormOfRankEq [Fintype ι] (b : Basis ι R M)
+    (h : Module.finrank R N = Module.finrank R M) :
+    Basis.SmithNormalForm N ι (Fintype.card ι) :=
+  let ⟨n, bM, bN, f, a, snf⟩ := N.smithNormalForm b
+  let e : Fin n ≃ Fin (Fintype.card ι) := Fintype.equivOfCardEq (by
+    simp only [Fintype.card_fin, ← Module.finrank_eq_card_basis bM, ← h,
+      Module.finrank_eq_card_basis bN])
+  ⟨bM, bN.reindex e, e.symm.toEmbedding.trans f, a ∘ e.symm, fun i ↦ by
+    simp only [snf, Basis.coe_reindex, Function.Embedding.trans_apply, Equiv.toEmbedding_apply,
+      (· ∘ ·)]⟩
+
+variable [Finite ι]
+
+/--
+If `M` is finite free over a PID `R`, then for any submodule `N` of the same rank,
+we can find basis for `M` and `N` with the same indexing such that the inclusion map
+is a square diagonal matrix.
+
+See also `Submodule.smithNormalFormOfRankEq` for a version of this theorem that returns
+a `Basis.SmithNormalForm`.
+-/
+theorem Submodule.exists_smith_normal_form_of_rank_eq (b : Basis ι R M)
+    (h : Module.finrank R N = Module.finrank R M) :
+    ∃ (b' : Basis ι R M) (a : ι → R) (ab' : Basis ι R N), ∀ i, (ab' i : M) = a i • b' i := by
+  cases nonempty_fintype ι
+  let ⟨bM, bN, f, a, snf⟩ := N.smithNormalFormOfRankEq b h
+  let e : Fin (Fintype.card ι) ≃ ι :=
+    Equiv.ofBijective f
+      ((Fintype.bijective_iff_injective_and_card f).mpr ⟨f.injective, Fintype.card_fin _⟩)
+  have fe : ∀ i, f (e.symm i) = i := e.apply_symm_apply
+  exact
+    ⟨bM, a ∘ e.symm, bN.reindex e, fun i ↦ by
+      simp only [snf, fe, Basis.map_apply, LinearEquiv.restrictScalars_apply R,
+          restrictScalarsEquiv_apply, Basis.coe_reindex, (· ∘ ·)]⟩
+
+/--
+If `M` is finite free over a PID `R`, then for any submodule `N` of the same rank,
+we can find basis for `M` and `N` with the same indexing such that the inclusion map
+is a square diagonal matrix; this is the basis for `M`.
+See `Submodule.smithNormalFormBotBasis` for the basis on `N`,
+see `Submodule.smithNormalFormCoeffs` for the entries of the diagonal matrix
+and `Submodule.smithNormalFormBotBasis_def` for the proof that the inclusion map
+forms a square diagonal matrix.
```

Comment:
Thanks for writing such detailed docstrings!
What do you think of formatting this as a markdown bullet list?
```suggestion
is a square diagonal matrix; this is the basis for `M`. See

* `Submodule.smithNormalFormBotBasis` for the basis on `N`,
* `Submodule.smithNormalFormCoeffs` for the entries of the diagonal matrix
* `Submodule.smithNormalFormBotBasis_def` for the proof that the inclusion map
  forms a square diagonal matrix.
```
and similarly for the docstrings below.

Reply:
Thanks for the review. I have modified the docs as suggested. For consistency, I did the same thing to the original docstrings that I copied and adapted.